From 62591a21a2937fb2907b923afaf5fd87484d8e95 Mon Sep 17 00:00:00 2001
From: Tatsuo ONISHI <onishi@latmos.ipsl.fr>
Date: Fri, 7 Oct 2022 10:47:00 +0200
Subject: [PATCH 27/52] Tracers (CHEM_TRACE2)

TYPE : Updates

list of modified files:
---------------------------------------------------------------------------------
chem/emissions_driver.F
  * NO and CO emissions are hacked
  * NO emissions are used to differentiate 20 tracer emissions.
      tracer 1 : NO = 1.0
      tracer 2 : NO = 2.0
      tracer 3 : NO = 3.0
      ....
      tracer 20: NO = 20.0
  * CO emissions are used to specify tracer emission intensities.

chem/module_input_chem_data.F
  * Set boundary values of tracers to 1.e-12

Registry/registry.chem
  * package chem_trace2 is updated.
---
 Registry/registry.chem        |  4 +-
 chem/emissions_driver.F       | 72 +++++++++++++++++++++++++++++++++++
 chem/module_input_chem_data.F | 33 ++++++++++++++--
 3 files changed, 104 insertions(+), 5 deletions(-)

diff --git a/Registry/registry.chem b/Registry/registry.chem
index b5cda202..62a9a0fd 100644
--- a/Registry/registry.chem
+++ b/Registry/registry.chem
@@ -4226,7 +4226,9 @@ package   cbmz_mosaic_8bin_aq  chem_opt==10          -             chem:so2,sulf
 package   radm2sorg_aq chem_opt==11                  -             chem:so2,sulf,no2,no,o3,hno3,h2o2,ald,hcho,op1,op2,paa,ora1,ora2,nh3,n2o5,no3,pan,hc3,hc5,hc8,eth,co,ol2,olt,oli,tol,xyl,aco3,tpan,hono,hno4,ket,gly,mgly,dcb,onit,csl,iso,hcl,ho,ho2,so4aj,so4ai,nh4aj,nh4ai,no3aj,no3ai,naaj,naai,claj,clai,orgaro1j,orgaro1i,orgaro2j,orgaro2i,orgalk1j,orgalk1i,orgole1j,orgole1i,orgba1j,orgba1i,orgba2j,orgba2i,orgba3j,orgba3i,orgba4j,orgba4i,orgpaj,orgpai,ecj,eci,p25j,p25i,antha,seas,soila,nu0,ac0,corn,so4cwj,so4cwi,nh4cwj,nh4cwi,no3cwj,no3cwi,nacwj,nacwi,clcwj,clcwi,orgaro1cwj,orgaro1cwi,orgaro2cwj,orgaro2cwi,orgalk1cwj,orgalk1cwi,orgole1cwj,orgole1cwi,orgba1cwj,orgba1cwi,orgba2cwj,orgba2cwi,orgba3cwj,orgba3cwi,orgba4cwj,orgba4cwi,orgpacwj,orgpacwi,eccwj,eccwi,p25cwj,p25cwi,anthcw,seascw,soilcw,nu0cw,ac0cw,corncw
 package   racmsorg_aq  chem_opt==12                  -             chem:so2,sulf,no2,no,o3,hno3,h2o2,ald,hcho,op1,op2,paa,ora1,ora2,nh3,n2o5,no3,pan,hc3,hc5,hc8,eth,co,ete,olt,oli,tol,xyl,aco3,tpan,hono,hno4,ket,gly,mgly,dcb,onit,csl,iso,co2,ch4,udd,hket,api,lim,dien,macr,ho,ho2,so4aj,so4ai,nh4aj,nh4ai,no3aj,no3ai,naaj,naai,claj,clai,orgaro1j,orgaro1i,orgaro2j,orgaro2i,orgalk1j,orgalk1i,orgole1j,orgole1i,orgba1j,orgba1i,orgba2j,orgba2i,orgba3j,orgba3i,orgba4j,orgba4i,orgpaj,orgpai,ecj,eci,p25j,p25i,antha,seas,soila,nu0,ac0,corn,so4cwj,so4cwi,nh4cwj,nh4cwi,no3cwj,no3cwi,nacwj,nacwi,clcwj,clcwi,orgaro1cwj,orgaro1cwi,orgaro2cwj,orgaro2cwi,orgalk1cwj,orgalk1cwi,orgole1cwj,orgole1cwi,orgba1cwj,orgba1cwi,orgba2cwj,orgba2cwi,orgba3cwj,orgba3cwi,orgba4cwj,orgba4cwi,orgpacwj,orgpacwi,eccwj,eccwi,p25cwj,p25cwi,anthcw,seascw,soilcw,nu0cw,ac0cw,corncw
 package   chem_tracer   chem_opt==13                 -             chem:so2,no,ald,hcho,ora2,co
-package   chem_trace2   chem_opt==14                 -             chem:tracer_1
+#jcr pour les traceurs
+# package   chem_trace2   chem_opt==14                 -             chem:tracer_1
+package   chem_trace2   chem_opt==14                 -             chem:tracer_1,tracer_2,tracer_3,tracer_4,tracer_5,tracer_6,tracer_7,tracer_8,tracer_9,tracer_10,tracer_11,tracer_12,tracer_13,tracer_14,tracer_15,tracer_16,tracer_17,tracer_18,tracer_19,tracer_20,tracer_ens
 package   chem_trace_ens   chem_opt==15              -             chem:tracer_1,tracer_2,tracer_3,tracer_4,tracer_5,tracer_6,tracer_7,tracer_8,tracer_9,tracer_10,tracer_11,tracer_12,tracer_13,tracer_14,tracer_15,tracer_16,tracer_17,tracer_18,tracer_19,tracer_20,tracer_ens
 
 # GHG chem packages, passive tracers
diff --git a/chem/emissions_driver.F b/chem/emissions_driver.F
index ae258bea..27e62e51 100644
--- a/chem/emissions_driver.F
+++ b/chem/emissions_driver.F
@@ -1233,6 +1233,78 @@ real, save :: freq_industry(24)    = &
              end do
           end do
        end do
+!jcr
+   CASE (CHEM_TRACE2)
+       do j=jts,jte
+          do i=its,ite
+             do k=kts,min(config_flags%kemit,kte-ksub)
+                conv = 4.828e-4/rho_phy(i,k,j)*dtstep/(dz8w(i,k,j)*60.)
+                IF(emis_ant(i,k,j,p_e_no).eq.1.0)                        &
+                chem(i,k,j,p_tracer_1)  = chem(i,k,j,p_tracer_1)         &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.2.0)                        &
+                chem(i,k,j,p_tracer_2)  = chem(i,k,j,p_tracer_2)         &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.3.0)                        &
+                chem(i,k,j,p_tracer_3)  = chem(i,k,j,p_tracer_3)         &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.4.0)                        &
+                chem(i,k,j,p_tracer_4)  = chem(i,k,j,p_tracer_4)         &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.5.0)                        &
+                chem(i,k,j,p_tracer_5)  = chem(i,k,j,p_tracer_5)         &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.6.0)                        &
+                chem(i,k,j,p_tracer_6)  = chem(i,k,j,p_tracer_6)         &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.7.0)                        &
+                chem(i,k,j,p_tracer_7)  = chem(i,k,j,p_tracer_7)         &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.8.0)                        &
+                chem(i,k,j,p_tracer_8)  = chem(i,k,j,p_tracer_8)         &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.9.0)                        &
+                chem(i,k,j,p_tracer_9)  = chem(i,k,j,p_tracer_9)         &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.10.0)                       &
+                chem(i,k,j,p_tracer_10)  = chem(i,k,j,p_tracer_10)       &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.11.0)                       &
+                chem(i,k,j,p_tracer_11)  = chem(i,k,j,p_tracer_11)       &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.12.0)                       &
+                chem(i,k,j,p_tracer_12)  = chem(i,k,j,p_tracer_12)       &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.13.0)                       &
+                chem(i,k,j,p_tracer_13)  = chem(i,k,j,p_tracer_13)       &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.14.0)                       &
+                chem(i,k,j,p_tracer_14)  = chem(i,k,j,p_tracer_14)       &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.15.0)                       &
+                chem(i,k,j,p_tracer_15)  = chem(i,k,j,p_tracer_15)       &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.16.0)                       &
+                chem(i,k,j,p_tracer_16)  = chem(i,k,j,p_tracer_16)       &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.17.0)                       &
+                chem(i,k,j,p_tracer_17)  = chem(i,k,j,p_tracer_17)       &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.18.0)                       &
+                chem(i,k,j,p_tracer_18)  = chem(i,k,j,p_tracer_18)       &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.19.0)                       &
+                chem(i,k,j,p_tracer_19)  = chem(i,k,j,p_tracer_19)       &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                IF(emis_ant(i,k,j,p_e_no).eq.20.0)                       &
+                chem(i,k,j,p_tracer_20)  = chem(i,k,j,p_tracer_20)       &
+                     +emis_ant(i,k,j,p_e_co)*conv
+                chem(i,k,j,p_tracer_ens)  = chem(i,k,j,p_tracer_ens)     &
+                     +emis_ant(i,k,j,p_e_co)*conv
+             end do
+          end do
+       end do
+!jcr
 !
     CASE(CO2_TRACER,GHG_TRACER)  ! for ghg_tracer package
 
diff --git a/chem/module_input_chem_data.F b/chem/module_input_chem_data.F
index edc7ed07..d4ea248b 100755
--- a/chem/module_input_chem_data.F
+++ b/chem/module_input_chem_data.F
@@ -737,7 +737,8 @@ SUBROUTINE input_chem_profile (si_grid)
 !     si_grid%chem(ims:ime,kms:kme,jms:jme,p_so2) = 0.0001
       si_grid%chem(ims:ime,kms:kme,jms:jme,p_co ) = 0.08
    ELSE IF ( si_grid%chem_opt == CHEM_TRACE2 ) THEN
-      si_grid%chem(ims:ime,kms:kme,jms:jme,p_TRACER_1 ) = 0.08
+!jcr       si_grid%chem(ims:ime,kms:kme,jms:jme,p_TRACER_1 ) = 0.08 \
+      si_grid%chem(ims:ime,kms:kme,jms:jme,p_TRACER_1 ) = 1.e-12
    ELSE IF ( si_grid%chem_opt == CHEM_VASH ) THEN
       si_grid%chem(ims:ime,kms:kme,jms:jme,1:numgas ) = 0.
    ELSE IF ( si_grid%chem_opt == CHEM_VOLC ) THEN
@@ -1317,9 +1318,33 @@ SUBROUTINE input_chem_profile (si_grid)
     else
        chem = conmin
     end if
-    if( nch .eq. p_tracer_1  ) then
-       chem = 0.08
-    endif
+!jcr    if( nch .eq. p_tracer_1  ) then
+!jcr       chem = 0.08
+!jcr    endif
+!jcr
+    if( nch .eq. p_tracer_1  ) chem = 1.e-12
+    if( nch .eq. p_tracer_2  ) chem = 1.e-12
+    if( nch .eq. p_tracer_3  ) chem = 1.e-12
+    if( nch .eq. p_tracer_4  ) chem = 1.e-12
+    if( nch .eq. p_tracer_5  ) chem = 1.e-12
+    if( nch .eq. p_tracer_6  ) chem = 1.e-12
+    if( nch .eq. p_tracer_7  ) chem = 1.e-12
+    if( nch .eq. p_tracer_8  ) chem = 1.e-12
+    if( nch .eq. p_tracer_9  ) chem = 1.e-12
+    if( nch .eq. p_tracer_10 ) chem = 1.e-12
+    if( nch .eq. p_tracer_11 ) chem = 1.e-12
+    if( nch .eq. p_tracer_12 ) chem = 1.e-12
+    if( nch .eq. p_tracer_13 ) chem = 1.e-12
+    if( nch .eq. p_tracer_14 ) chem = 1.e-12
+    if( nch .eq. p_tracer_15 ) chem = 1.e-12
+    if( nch .eq. p_tracer_16 ) chem = 1.e-12
+    if( nch .eq. p_tracer_17 ) chem = 1.e-12
+    if( nch .eq. p_tracer_18 ) chem = 1.e-12
+    if( nch .eq. p_tracer_19 ) chem = 1.e-12
+    if( nch .eq. p_tracer_20 ) chem = 1.e-12
+    if( nch .eq. p_tracer_ens ) chem = 1.e-12
+!jcr
+
 
   END SUBROUTINE bdy_chem_value_tracer
 !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
-- 
2.31.1

